modules/flann/include/opencv2/flann/heap.h

```diff
@@ -49,117 +61,174 @@ namespace cvflann
      */
     bool empty()
     {
-        return size()==0;
+        return heap.empty();
     }
 
     /**
-     * Clears the heap.
+     * \brief Clears the heap.
      */
     void clear()
     {
         heap.clear();
-        count = 0;
     }
 
-    struct CompareT
+    /**
+     * \brief Sets the heap maximum capacity.
+     *
+     * \param capacity heap maximum capacity
+     */
+    void reserve(const int capacity)
     {
-        bool operator()(const T& t_1, const T& t_2) const
-        {
-            return t_2 < t_1;
-        }
-    };
+        heap.reserve(capacity);
+    }
 
     /**
-     * Insert a new element in the heap.
+     * \brief Inserts a new element in the heap.
      *
      * We select the next empty leaf node, and then keep moving any larger
      * parents down until the right location is found to store this element.
      *
-     * Params:
-     *     value = the new element to be inserted in the heap
+     * \param value the new element to be inserted in the heap
      */
     void insert(T value)
     {
         /* If heap is full, then return without adding this element. */
-        if (count == length) {
+        if (size() == capacity()) {
             return;
         }
 
         heap.push_back(value);
-        static CompareT compareT;
-        std::push_heap(heap.begin(), heap.end(), compareT);
-        ++count;
+        std::push_heap(heap.begin(), heap.end(), greater<T>());
     }
 
-
-
     /**
-     * Returns the node of minimum value from the heap (top of the heap).
+     * \brief Returns the node of minimum value from the heap (top of the heap).
      *
-     * Params:
-     *     value = out parameter used to return the min element
-     * Returns: false if heap empty
+     * \param[out] value parameter used to return the min element
+     * \returns false if heap empty
      */
     bool popMin(T& value)
     {
-        if (count == 0) {
+        if (empty()) {
             return false;
         }
 
         value = heap[0];
-        static CompareT compareT;
-        std::pop_heap(heap.begin(), heap.end(), compareT);
+        std::pop_heap(heap.begin(), heap.end(), greater<T>());
         heap.pop_back();
-        --count;
 
         return true;  /* Return old last node. */
     }
+
+    /**
+     * \brief Returns a shared heap for the given memory pool ID.
+     *
+     * It constructs the heap if it does not already exists.
+     *
+     * \param poolId a user-chosen hashable ID for identifying the heap.
+     *     For thread-safe operations, using current thread ID is a good choice.
+     * \param capacity heap maximum capacity
+     * \param iterThreshold remove heaps that were not reused for more than specified iterations count
+     * \returns pointer to the heap
+     */
+    template <typename HashableT>
+    static cv::Ptr<Heap<T>> getPooledInstance(
+        const HashableT& poolId, const int capacity, const int iterThreshold = 2 * cv::getNumThreads())
```

Comment:
> = 2 * cv::getNumThreads()

It makes sense to calculate such values internally.
E.g use `= 0` in parameters list and replace that value later (value should be >= 2)